runtime/lua/vim/treesitter.lua

```diff
@@ -94,25 +94,43 @@ function M.get_parser(bufnr, lang, opts)
 
   if not valid_lang(lang) then
     if not parsers[bufnr] then
-      error(
-        string.format(
-          'There is no parser available for buffer %d and one could not be'
-            .. ' created because lang could not be determined. Either pass lang'
-            .. ' or set the buffer filetype',
-          bufnr
-        )
-      )
+      return nil
     end
   elseif parsers[bufnr] == nil or parsers[bufnr]:lang() ~= lang then
-    assert(lang, 'lang should be valid')
-    parsers[bufnr] = M._create_parser(bufnr, lang, opts)
+    local parser = vim.F.npcall(M._create_parser, bufnr, lang, opts)
+    if not parser then
+      return nil
+    end
+    parsers[bufnr] = parser
   end
 
   parsers[bufnr]:register_cbs(opts.buf_attach_cbs)
 
   return parsers[bufnr]
 end
 
+--- Returns the parser for a specific buffer and attaches it to the buffer
+---
+--- If needed, this will create the parser.
+---
+---@param bufnr (integer|nil) Buffer the parser should be tied to (default: current buffer)
+---@param lang (string|nil) Language of this parser (default: from buffer filetype)
+---@param opts (table|nil) Options to pass to the created language tree
+---
+---@return vim.treesitter.LanguageTree object to use for parsing
+function M.get_parser(bufnr, lang, opts)
+  -- TODO(ribru17): Remove _get_parser and move that logic back here once the breaking function
+  -- signature change is acceptable.
+  local parser = M._get_parser(bufnr, lang, opts)
+  if not parser then
+    error(
```

Comment:
I think such messages should be `notify_once` (like `vim.deprecate`) to avoid spam (in addition to a simple `error` to trip `pcall`). We should also be 1. more concrete (specify versions) and 2. more concise (try to keep messages to a single line to avoid "press enter")

Reply:
My guess is that non-trivial amount of usages rely on `pcall` to test if there is a parser (like in `vim._comment`). So this error will be not visible until "rip the band-aid off" stage of deprecation.

It's just an observation though, as having this is better than immediate behavior change and I don't have better suggestions (apart from trying to advertise this as much as reasonably possible).